smarts/sstudio/genhistories.py

```diff
@@ -54,14 +62,82 @@ def __init__(self, dataset_spec: Dict[str, Any], output: str):
         self._log = logging.getLogger(self.__class__.__name__)
         self.check_dataset_spec(dataset_spec)
         self._output = output
-        self._path = dataset_spec["input_path"]
+        self._path = os.path.expanduser(dataset_spec["input_path"])
         real_lane_width_m = dataset_spec.get("real_lane_width_m", DEFAULT_LANE_WIDTH)
-        lane_width = dataset_spec.get("map_net", {}).get(
-            "lane_width", real_lane_width_m
-        )
+        lane_width = self._map_spec.get("lane_width", real_lane_width_m)
         self._scale = lane_width / real_lane_width_m
         self._flip_y = dataset_spec.get("flip_y", False)
         self._swap_xy = dataset_spec.get("swap_xy", False)
+        # most trajectory datasets have .1s time delta (i.e., were collected at 10 Hz)
+        self._dt_sec = 0.1
+
+    class _WindowedReader:
+        """Iterates over the rows in file using a sliding window that keeps track of
+        both a number of rows before the current row and a number of rows after.
+        These "windows" are passed to a row transformation function on each step.
+        For example, if window_before = 4 and window_after = 3, for a 9-row
+        file, the windows associated with each row are:
+            row 1, before = [], after = [2, 3, 4]
+            row 2, before = [1], after = [3, 4, 5]
+            row 3, before = [2, 1], after = [4, 5, 6]
+            row 4, before = [3, 2, 1], after = [5, 6, 7]
+            row 5, before = [4, 3, 2, 1], after = [6, 7, 8]
+            row 6, before = [5, 4, 3, 2], after = [7, 8, 9]
+            row 7, before = [6, 5, 4, 3], after = [8, 9]
+            row 8, before = [7, 6, 5, 4], after = [9]
+            row 9, before = [8, 7, 6, 5], after = []
+        Windows are cleared whenever the value in the (optional) `group_col` column changes.
+        This was designed to be nestable by making the `row_gen` parameter support an iterator over another _WindowedReader.
+        """
+
+        Row = Dict[str, Any]
+
+        def __init__(
+            self,
+            row_gen: Iterable[Row],
+            transform_fn: Callable[[Row, Deque[Row], Deque[Row]], None],
+            window_before: int = 0,
+            window_after: int = 0,
+            group_col: Optional[str] = None,
+        ):
+            self._row_gen = row_gen
+            self._transform_fn = transform_fn
+            self._before_width = window_before
+            self._after_width = window_after
```

Comment:
Should there be a default value of these instead of 0 or they vary for different map source?

Reply:
Yeah, I think it varies by how much "smoothing" the user wants to do within a window.  A default of 0 is basically no smoothing.  (So, for example, because the NGSIM dataset is noisy, we currently smooth its positions with a window size of 15 -- 7 before and 7 after the current row.  But there are circumstances, such as the contest perhaps, where we might just want to use the positions "as is", which would mean 0 before and 0 after the current row.)